docs/source/cookbook/clients.rst

```diff
@@ -20,30 +20,38 @@ Code
 
     #! /usr/bin/env python
 
+    import sys
+    try:
+        from cStringIO import StringIO as BytesIO
+    except ImportError:
+        from io import BytesIO
+
     from ldaptor.protocols.ldap.ldapclient import LDAPClient
     from ldaptor.protocols.ldap.ldapsyntax import LDAPEntry
-    from twisted.internet.defer import inlineCallbacks, returnValue
+    from twisted.internet.defer import inlineCallbacks
     from twisted.internet.endpoints import clientFromString, connectProtocol
     from twisted.internet.task import react
     from twisted.python import log
-    from cStringIO import StringIO
-    import sys
+
 
     @inlineCallbacks
     def onConnect(clientProtocol):
         o = LDAPEntry(clientProtocol, "dc=org")
         resultList = yield o.search()
-        f = StringIO()
+        f = BytesIO()
         for result in resultList:
-            f.write(str(result))
-            f.write("\n")
-        log.msg("LDIF formatted results:\n{}".format(f.getvalue()))
+            f.write(result.toWire())
```

Comment:
That's right, we need some method (in entry classes at least) which will output its LDIF representation as unicode. Method `asLDIF` is already used in `delta` classes but it returns bytes representation. `__str__` method could be appropriate but I'm afraid it could be misused as it had `toWire` functionality in the past. Maybe `getText`? It can be implemented in another pull request.

I found out that `LDAPException.__str__` method was used to get error bytes representation and send it to LDAP client in case of error. This approach could not work in Python 3 as exception message must be unicode, so I implemented the same protocol as for `pureber` and `pureldap` module classes. It looks like some of these exceptions are not sent to LDAP clients but actually _raised_ so they should be allowed to get their human readable messages without `DeprecationWarning` :)

> Like for the exception, add a `Exception.message` member and use that.

> But I see that here we have message...but the string representation is something else...and I don't understand why message content is resolved only on access and not when set, in init and in setter?

True, I'll fix it in another PR.

> and for type and value it looks like `toWire` is used with the semantic of `toPreWire` as the output is used in another `toWire` which does the actual BER encoding.
> For the `LDAPAttributeTypeAndValue` we should have something like `getAttribute()` which will return text/Unicode and can be later decoded in the `toWire()` method of the caller code.

Maybe it is more straightforward if all internal LDAP objects have a single method name which returns their bytes representation? Or the emphasis here is that bytes representation of `distinguishedname` objects makes little sense and they should return their values as unicode?

Reply:
> Or the emphasis here is that bytes representation of `distinguishedname` objects makes little sense and they should return their values as unicode?

Yes. For DN and any other key/value attributes  we can have a `getText` method.
I can't think of a better name.

I expect that some attributes will also have a binary value, so we can also have getBytes() or getComponents() a similar method... as for a binary value you might encode the key as UTF-8 and value as other name.

Reply:
Thanks! I think I will fix it in one of the nearest pull requests.